### PR TITLE
Add utility function to parse details from S3 URIs and filenames

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Create a central utility function to parse TIMDEX "run" details from names of files that are used/created throughout the Transmogrifier A/B Diff workflow. This function replaces the now deprecated `parse_x` functions that were previously implemented within core function modules. Here are the main updates: 

* Utility function [utils.parse_timdex_filename](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/42/files#diff-3257f756c2ef8ad26bddac437d0041e415592ff08a5e6618f1b3b52e369f0259R82-R102)
* Use utility function output to create [`get_transformed_filename` function](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/42/files#diff-7a49d9dbf619fca67eb5ed5ed16b6c62440a4cc6898754cd835d2b819be74ddcR314-R323) in `run_ab_transforms`.
* Use utility function output to [name columns in the collated dataset](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/42/files#diff-1cf5419c409425ac3b8275acb901b7c68cc1131a5b24d5846f477eb967681e55R95-R103) in `collate_ab_transforms`.

### How can a reviewer manually see the effects of these changes?
Run `make test` and confirm all unit tests are passing.

### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TIMX-380
- https://mitlibraries.atlassian.net/browse/TIMX-365

### Developer
- [X] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

**Comment by jonavellecuerdo** [comment]
@ehanson8 This is a small, "cleanup" body of work that @ghukill and I wanted to implement. Please feel free to take a look if you're interested!

File: tests/test_utils.py [hunk 1]
```diff
@@ -56,6 +57,53 @@ def test_create_sub_directories_success(tmp_path):
...
+        "source-2024-01-01-full-extracted-records-to-index.xml"
+    ) == {
+        "source": "source",
+        "date": "2024-01-01",
+        "cadence": "full",
+        "stage": "extracted",
+        "action": "index",
+        "index": None,
+        "file_type": "xml",
+    }
```
Comment (ghukill) [inline tests/test_utils.py:99]: These kind of simple tests are so nice, give confidence it works as expected.

**Review by ghukill — APPROVED** [review]
Not much to comment on, looks good and works as expected!

File: abdiff/core/utils.py [hunk 2]
```diff
@@ -78,6 +79,29 @@ def load_dataset(base_dir: str | Path) -> ds.Dataset:
...
+def parse_timdex_filename(s3_uri_or_filename: str) -> dict[str, str | None]:
+    """Parse details from filename."""
+    filename = s3_uri_or_filename.split("/")[-1]
+
+    match_result = re.match(
+        r"^([\w\-]+?)-(\d{4}-\d{2}-\d{2})-(\w+)-(\w+)-records-to-(.+?)(?:_(\d+))?\.(\w+)$",
+        filename,
+    )
+
+    keys = ["source", "date", "cadence", "stage", "action", "index", "file_type"]
```
Comment (ghukill) [inline abdiff/core/utils.py:91]: Apologies!  Missed this on first pass.

Can we update these now, as discussed we might do in the TIMDEX parquet work? 

- `date` --> `run-date`
- `cadence` --> `run-type`

**Review by ghukill — CHANGES_REQUESTED** [review]
Requesting one small change, `date` to `run-date` and `cadence` to `run-type` as discussed for the TIMDEX parquet work.

Otherwise, looks good to me!  Approval still stands with this change.

File: abdiff/core/utils.py (hunk 2 above)
Comment (ghukill) [reply abdiff/core/utils.py:91]: Here is a comment from the engineering plan: https://mitlibraries.atlassian.net/wiki/spaces/IN/pages/edit-v2/4094296066?focusedCommentId=4141219863.

The name change is not performed yet in this engineering plan, but I think it was a good idea and will make that change.

Comment (jonavellecuerdo) [reply abdiff/core/utils.py]: Good catch! Made the required updates. See latest commit [9425ce3](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/42/commits/9425ce3b4a3c44267944687387623b73bfc3af67). 🤓

File: abdiff/core/run_ab_transforms.py [hunk 3]
```diff
@@ -310,22 +311,20 @@ def validate_output(
...
+    output_filename = (
+        "{source}-{run_date}-{run_type}-{stage}-records-to-index{index}.{file_type}"
+    )
+    return output_filename.format(
+        source=filename_details["source"],
+        run_date=filename_details["run-date"],
+        run_type=filename_details["run-type"],
+        stage=filename_details["stage"],
+        index=filename_details["index"],
+        file_type=filename_details["file_type"],
```
Comment (ghukill) [inline abdiff/core/run_ab_transforms.py:329]: Ack... sorry the dash vs underscore required this... but seems okay.

**Pull request merged**